### PR TITLE
SONAR-7436 monitor CE process

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 import java.io.File;
 import org.slf4j.LoggerFactory;
+import sun.misc.VMSupport;
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java]: MAJOR Replace this usage of Sun classes by ones from the Java API. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1191)

File: server/sonar-process/src/main/java/org/sonar/process/jmx/EsSettingsMBean.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process.jmx;
+
+/**
+ * MBean registered in the Elasticsearch process
+ */
+public interface EsSettingsMBean {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/EsSettingsMBean.java]: MINOR Move constants to a class or enum. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1214)

File: server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeDatabaseMonitor.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    this.jmxConnector = jmxConnector;
+  }
+
+  @Override
+  public String name() {
+    return "Compute Engine Database";
+  }
+
+  @Override
+  public LinkedHashMap<String, Object> attributes() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeDatabaseMonitor.java]: MAJOR The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

File: server/sonar-server/src/test/java/org/sonar/server/computation/queue/CeQueueInitializerTest.java [hunk 4]
```diff
@@ -53,30 +45,9 @@
   public TemporaryFolder tempFolder = new TemporaryFolder();
```
Comment (SonarTech) [inline server/sonar-server/src/test/java/org/sonar/server/computation/queue/CeQueueInitializerTest.java]: MAJOR Remove this unused "TemporaryFolder". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2924)

File: server/sonar-process/src/main/java/org/sonar/process/jmx/CeTasksMBean.java [hunk 5]
```diff
@@ -17,14 +17,11 @@
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-package org.sonar.server.computation.monitoring;
+package org.sonar.process.jmx;
 
-public interface ComputeEngineQueueMonitorMBean {
+public interface CeTasksMBean {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/CeTasksMBean.java]: MINOR Move constants to a class or enum. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1214)

File: server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeTasksMonitor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    this.jmxConnector = jmxConnector;
+  }
+
+  @Override
+  public String name() {
+    return "Compute Engine Tasks";
+  }
+
+  @Override
+  public LinkedHashMap<String, Object> attributes() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeTasksMonitor.java]: MAJOR The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

File: server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeStateMonitor.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    this.jmxConnector = jmxConnector;
+  }
+
+  @Override
+  public String name() {
+    return "Compute Engine State";
+  }
+
+  @Override
+  public LinkedHashMap<String, Object> attributes() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/CeStateMonitor.java]: MAJOR The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

File: server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/EsStateMonitor.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    this.jmxConnector = jmxConnector;
+  }
+
+  @Override
+  public String name() {
+    return "Elasticsearch State";
+  }
+
+  @Override
+  public LinkedHashMap<String, Object> attributes() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/EsStateMonitor.java]: MAJOR The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap". [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

File: server/sonar-process/src/main/java/org/sonar/process/jmx/CeDatabaseMBean.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process.jmx;
+
+public interface CeDatabaseMBean {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/CeDatabaseMBean.java]: MINOR Move constants to a class or enum. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1214)

File: server/sonar-process/src/main/java/org/sonar/process/AllProcessesCommands.java [hunk 10]
```diff
@@ -22,10 +22,15 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
 import java.nio.MappedByteBuffer;
 import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
 import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/AllProcessesCommands.java]: MINOR Remove this unused import 'org.apache.commons.lang.StringUtils'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java [hunk 11]
```diff
@@ -88,6 +90,15 @@ private ProcessBuilder create(JavaCommand javaCommand) {
     return processBuilder;
   }
 
+  /**
+   * JVM option to enable the agent that allows inter-process communication through JMX without
+   * opening new ports. The agent is available in JRE of OpenJDK/OracleJDK only.
+   * @see ProcessEntryPoint
+   */
+  private String getJmxAgentCommand() {
```
Comment (SonarTech) [inline server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java]: MAJOR Make "getJmxAgentCommand" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 13 issues:
* MAJOR 9 major
* MINOR 4 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Make "configureIndexDefaults" a "static" method.](https://github.com/SonarSource/sonarqube/blob/30303896c3f741b3f7605f221372427fa96a7eb7/server/sonar-search/src/main/java/org/sonar/search/EsSettings.java#L149) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/30303896c3f741b3f7605f221372427fa96a7eb7/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

**Comment by SonarTech** [comment]
SonarQube analysis reported 13 issues:
* MAJOR 9 major
* MINOR 4 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Make "configureIndexDefaults" a "static" method.](https://github.com/SonarSource/sonarqube/blob/559cb1682dd4fcfe6e1dcbe2b014e230e7938e69/server/sonar-search/src/main/java/org/sonar/search/EsSettings.java#L149) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/559cb1682dd4fcfe6e1dcbe2b014e230e7938e69/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* MAJOR 6 major
* MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/dd0344fef2c2ff8e0358132cef63ba13be27e12d/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

File: server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java [hunk 12]
```diff
@@ -79,7 +81,7 @@ public void real_start() throws IOException {
         CONTAINER_ITSELF
           + 77 // level 4
           + 5 // content of CeModule
-          + 7 // content of CeQueueModule
+          + 8 // content of CeQueueModule
```
Comment (sns-seb) [inline server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java]: number of components in `CeQueueModule` is still 7, why this change?

File: server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java [hunk 13]
```diff
@@ -33,18 +33,19 @@
 import org.sonar.process.ProcessEntryPoint;
 import org.sonar.process.ProcessUtils;
 
-public class JavaProcessLauncher {
+class JavaProcessLauncher implements AutoCloseable {
```
Comment (sns-seb) [inline server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java]: why adding this interface? I don't see it used anywhere

File: server/sonar-process/src/main/java/org/sonar/process/ProcessId.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process;
+
+public enum ProcessId {
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessId.java]: This enum is a nice idea.
It would be neat to have it used everywhere.
When you rebased this branch on master, I don't think you updated master's code to use it (eg. https://github.com/SonarSource/sonarqube/blob/master/server/sonar-ce/src/main/java/org/sonar/ce/app/WebServerWatcherImpl.java#L42)

File: server/sonar-process/src/main/java/org/sonar/process/jmx/Jmx.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      Class<?>[] interfaces = instance.getClass().getInterfaces();
+      for (Class<?> anInterface : interfaces) {
+        if (anInterface.getName().endsWith("MBean")) {
+          mbeanInterface = anInterface;
+          break;
+        }
+      }
+      if (mbeanInterface == null) {
+        throw new IllegalArgumentException("Can not find the MBean interface of class " + instance.getClass().getName());
+      }
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/Jmx.java]: lines 51 to 61 should be in a method and inline comment the javadoc of this method

File: server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+import org.sonar.process.DefaultProcessCommands;
+import org.sonar.process.ProcessEntryPoint;
+import org.sonar.process.ProcessId;
+import org.sonar.process.Props;
+
+/**
+ * Connects to JMX of other JVM processes
+ */
+@Immutable
+public class JmxConnector {
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: From the implementation, it seems this is class should rather be named `JmxConnectionFactory`

more classic of a naming, but more explicite imho

File: server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+  public JmxConnector(File ipcSharedDir) {
+    this.ipcSharedDir = ipcSharedDir;
+  }
+
+  public JmxConnector(Props props) {
+    this.ipcSharedDir = props.nonNullValueAsFile(ProcessEntryPoint.PROPERTY_SHARED_PATH);
+  }
+
+  public JmxConnection connect(ProcessId processId) {
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: this function could be renamed `create`

Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: especially because it will always return a new object (it's not a pool)

File: server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java (hunk 16 above)
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: side benefit, it would save the almost name conflict with `javax.management.remote.JMXConnector`

File: server/sonar-process/src/test/java/org/sonar/process/ProcessIdTest.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ProcessIdTest {
+
+  @Test
+  public void test() {
+    assertThat(ProcessId.COMPUTE_ENGINE.getKey()).isEqualTo("ce");
+    assertThat(ProcessId.COMPUTE_ENGINE.getIpcIndex()).isEqualTo(3);
```
Comment (sns-seb) [inline server/sonar-process/src/test/java/org/sonar/process/ProcessIdTest.java]: would be nice to have this kind of assertion for all values of the enum. Will enforce no one changes these values by mistake

File: server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnection.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+      return JMX.newMBeanProxy(connection, new ObjectName(mBeanName), mBeanInterfaceClass);
+    } catch (Exception e) {
+      throw new IllegalStateException("Fail to connect to MBean " + mBeanName, e);
+    }
+  }
+
+  public SortedMap<String, Object> getSystemState() {
+    SortedMap<String, Object> props = new TreeMap<>();
+    MemoryMXBean memory = getMBean(ManagementFactory.MEMORY_MXBEAN_NAME, MemoryMXBean.class);
+    props.put("Heap Committed (MB)", toMegaBytes(memory.getHeapMemoryUsage().getCommitted()));
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnection.java]: you should store `memory.getHeapMemoryUsage()` and `memory.getNonHeapMemoryUsage()` into a variable to be sure you get consistent values + that would make sure there is no performance issue if these methods do some IOs

File: server/sonar-search/src/main/java/org/sonar/search/EsSettings.java [hunk 20]
```diff
@@ -156,10 +172,10 @@ private void configureCluster(ImmutableSettings.Builder builder) {
       }
     }
     builder.put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, replicationFactor);
-    builder.put("cluster.name", props.value(ProcessProperties.CLUSTER_NAME));
+    builder.put("cluster.name", getClusterName());
     builder.put("cluster.routing.allocation.awareness.attributes", "rack_id");
-    builder.put("node.rack_id", props.value(ProcessProperties.CLUSTER_NODE_NAME, "unknown"));
-    builder.put("node.name", props.value(ProcessProperties.CLUSTER_NODE_NAME));
+    builder.put("node.rack_id", getNodeName());
+    builder.put("node.name", getNodeName());
```
Comment (sns-seb) [inline server/sonar-search/src/main/java/org/sonar/search/EsSettings.java]: you should use a variable to avoid risk of getting inconsistent values in name and rack_id (and the extra cost of call `props.value(...)`

File: server/sonar-server/src/main/java/org/sonar/ce/monitoring/CEQueueStatus.java [hunk 21]
```diff
@@ -134,5 +84,5 @@
   /**
    * Time spent processing batch reports since startup.
    */
-  long getProcessingTime();
+  long getProcessingTimeInMs();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/ce/monitoring/CEQueueStatus.java]: I don't think the `InMs` is useful. Is `Date.getTime()` named `Date.getTimeInMs()` ?
In addition, `addError` and `addSuccess` take a long which is a time in ms

File: server/sonar-server/src/main/java/org/sonar/server/computation/monitoring/CEQueueStatusImpl.java [hunk 22]
```diff
@@ -21,86 +21,53 @@
...
-  public long initPendingCount(long initialPendingCount) {
-    checkArgument(initialPendingCount >= 0, "Initial pending count must be >= 0");
-    checkState(
-      pending.compareAndSet(PENDING_INITIAL_VALUE, initialPendingCount),
-      "Method initPendingCount must be used before any other method and can not be called twice");
-    return initialPendingCount;
-  }
-
-  @Override
-  public long addReceived() {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/monitoring/CEQueueStatusImpl.java]: is the received metric good all together?
I think it is a valuable metric, otherwise we can't track the activity of the task submission at all anymore
obviously it should be on the WebServer side only now

File: server/sonar-server/src/main/java/org/sonar/server/computation/monitoring/CEQueueStatusImpl.java [hunk 23]
```diff
@@ -21,86 +21,53 @@
...
 
-  private void addProcessingTime(long time) {
-    checkArgument(time >= 0, "Processing time can not be < 0");
-    processingTime.addAndGet(time);
-  }
-
-  @Override
-  public long getReceivedCount() {
-    return received.get();
+  private void addProcessingTimeInMs(long ms) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/monitoring/CEQueueStatusImpl.java]: no value for `InMs` in name of this method, all long params in this class are milliseconds

File: server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/EsMonitor.java [hunk 24]
```diff
@@ -96,8 +109,8 @@ public int getNumberOfNodes() {
     for (Map.Entry<String, NodeStats> entry : nodesStats.getNodesMap().entrySet()) {
 
       LinkedHashMap<String, Object> nodeAttributes = new LinkedHashMap<>();
-      nodes.put(entry.getKey(), nodeAttributes);
       NodeStats stats = entry.getValue();
+      nodes.put(stats.getNode().getName(), nodeAttributes);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/EsMonitor.java]: `stats.getNode()` should be stored in a variable to have consistent values and avoid potential perf overhead

File: sonar-application/src/main/java/org/sonar/application/App.java [hunk 25]
```diff
@@ -38,15 +39,10 @@
...
-  public static final int APP_PROCESS_NUMBER = 0;
-  public static final int ES_PROCESS_INDEX = 1;
-  public static final int WEBSERVER_PROCESS_INDEX = 2;
-  public static final int CESERVER_PROCESS_INDEX = 3;
-
   private final Monitor monitor;
 
   public App(AppFileSystem appFileSystem, boolean watchForHardStop) {
-    this(Monitor.create(APP_PROCESS_NUMBER, appFileSystem, watchForHardStop));
+    this(Monitor.create(ProcessId.APP.getIpcIndex(), appFileSystem, watchForHardStop));
```
Comment (sns-seb) [inline sonar-application/src/main/java/org/sonar/application/App.java]: APP could be statically imported, so are ELASTICSEARCH and WEB_SERVER

File: server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java [hunk 26]
```diff
@@ -65,6 +66,7 @@ public void real_start() throws IOException {
     properties.setProperty(PATH_HOME, homeDir.getAbsolutePath());
     properties.setProperty(PATH_DATA, dataDir.getAbsolutePath());
     properties.setProperty(PATH_TEMP, tmpDir.getAbsolutePath());
+    properties.setProperty(ProcessEntryPoint.PROPERTY_SHARED_PATH, tmpDir.getAbsolutePath());
```
Comment (sns-seb) [inline server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java]: PROPERTY_SHARED_PATH could be statically imported

File: server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java [hunk 27]
```diff
@@ -104,12 +115,12 @@ private File buildPropertiesFile(JavaCommand javaCommand) {
       propertiesFile = File.createTempFile("sq-process", "properties");
       Properties props = new Properties();
       props.putAll(javaCommand.getArguments());
-      props.setProperty(ProcessEntryPoint.PROPERTY_PROCESS_KEY, javaCommand.getKey());
-      props.setProperty(ProcessEntryPoint.PROPERTY_PROCESS_INDEX, Integer.toString(javaCommand.getProcessIndex()));
+      props.setProperty(ProcessEntryPoint.PROPERTY_PROCESS_KEY, javaCommand.getProcessId().getKey());
```
Comment (sns-seb) [inline server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java]: `ProcessEntryPoint.*` could be statically imported

File: server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java [hunk 28]
```diff
@@ -104,6 +105,13 @@ public void launch(Monitored mp) {
         Thread.sleep(20L);
       }
 
+      // this property is set by the agent management-agent.jar enabled by org.sonar.process.monitor.JavaProcessLauncher.
+      String jmxUrl = VMSupport.getAgentProperties().getProperty("com.sun.management.jmxremote.localConnectorAddress");
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java]: this block could be in a method

File: server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java [hunk 29]
```diff
@@ -104,6 +105,13 @@ public void launch(Monitored mp) {
         Thread.sleep(20L);
       }
```
Comment (sns-seb) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java]: this block could be in a method
(I should have done it already, missed/forgot to do it)

**Comment by sns-seb** [comment]
many comments but only 1 functional one: missing received metric in the WebServer

File: server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java (hunk 13 above)
Comment (simonbrandhof) [inline server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/JavaProcessLauncher.java]: indeed no usage, but why not adding the interface ? It sounds to be a good practice.

File: server/sonar-process/src/main/java/org/sonar/process/ProcessId.java (hunk 14 above)
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessId.java]: Indeed. Good catch.

File: server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java (hunk 16 above)
Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: in this case the method is not `connect()` but `create`. Sounds good.

Comment (simonbrandhof) [inline server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnector.java]: and it requires to rename class `JmxConnectorProvider` to `JmxConnectionFactoryProvider`. Spring are you here ?

File: server/sonar-server/src/main/java/org/sonar/ce/monitoring/CEQueueStatus.java (hunk 21 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/ce/monitoring/CEQueueStatus.java]: This pattern helped me a lot in the past to avoid bugs. I don't think you can compare with java.util.Date#getTime(), which is a central piece of the JDK.

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/ce/monitoring/CEQueueStatus.java]: What possible bug?
And what else could it be but milliseconds anyway?

"It's a time! It's a long! It's milliseconds!"

All other longs in this interface are milliseconds and there is no need to specify it anywhere else than in the Javadoc. Longs as milliseconds in (parameters), longs as milliseconds out.

File: server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java [hunk 30]
```diff
@@ -118,6 +121,15 @@ public void launch(Monitored mp) {
     }
   }
 
+  private String guessJmxUrl() {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java]: MAJOR Make "guessJmxUrl" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues:
* MAJOR 7 major
* MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/1826c05e83be088f7bd4372499a7c419d05a8dbf/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues:
* MAJOR 7 major
* MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/8379ed298ca278553aa9d55232cb8408639d1767/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* MAJOR 6 major
* MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The return type of this method should be an interface such as "Map" rather than the implementation "LinkedHashMap".](https://github.com/SonarSource/sonarqube/blob/e840e426fb152fd5596c1a06d3c3260ed8d2b994/server/sonar-server/src/main/java/org/sonar/server/platform/monitoring/JvmPropsMonitor.java#L35) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1319)

**Comment by simonbrandhof** [comment]
Feedback applied. The monitoring metric `received` will come back in another PR.
